p2p/server/msgserver.go

```diff
@@ -147,27 +148,37 @@ func (p *MessageServer) handleMessage(msg Message) {
 }
 
 func (p *MessageServer) handleRequestMessage(msg Message, data *service.DataMsgWrapper) {
-	if payload := p.msgRequestHandlers[MessageType(data.MsgType)](msg); payload != nil {
-		rmsg := &service.DataMsgWrapper{MsgType: data.MsgType, ReqID: data.ReqID, Payload: payload}
-		sendErr := p.network.SendWrappedMessage(msg.Sender(), p.name, rmsg)
-		if sendErr != nil {
-			p.Error("Error sending response message, err:", sendErr)
+	p.Debug("handleRequestMessage start")
+	if foo, okFoo := p.msgRequestHandlers[MessageType(data.MsgType)]; okFoo {
+		if payload := foo(msg); payload != nil {
+			p.Debug("handle request type %v", data.MsgType)
+			rmsg := &service.DataMsgWrapper{MsgType: data.MsgType, ReqID: data.ReqID, Payload: payload}
+			sendErr := p.network.SendWrappedMessage(msg.Sender(), p.name, rmsg)
+			if sendErr != nil {
+				p.Error("Error sending response message, err:", sendErr)
+			}
+		} else {
+			p.Error("handler returned this request type: %v payload %v", data.MsgType, msg.Data())
 		}
+	} else {
+		p.Error("handler missing for request %v payload %v", data.ReqID)
 	}
+	p.Debug("handleRequestMessage close")
 }
 
 func (p *MessageServer) handleResponseMessage(headers *service.DataMsgWrapper) {
 	//get and remove from pendingMap
-	p.pendMutex.Lock()
+	p.Debug("handleResponseMessage req %v", headers.ReqID)
+	p.pendMutex.RLock()
 	foo, okFoo := p.resHandlers[headers.ReqID]
-	p.pendMutex.Unlock()
+	p.pendMutex.RUnlock()
```

Comment:
just a note for the future, maybe we can make response handlers static (like requests) and then we won't have to append them, protect the map and run a different callback every time.

Reply:
good idea!